### PR TITLE
WiP: Added azblob adapter

### PR DESCRIPTION
**Pull request by dim**
@andyborn @jpmedforth @mxmCherry @vrih

~~There are some breaking changes with this one that will require a major release (will comment inline)~~

### CODE REVIEW

File: bfsaz/bfsaz.go [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+		}
+		return pe
+	}
+
+	return err
+}
+
+func transKeys(meta map[string]string, from, to string) map[string]string {
+	for k, v := range meta {
+		if nk := strings.ReplaceAll(k, from, to); nk != k {
```
Comment (mxmCherry) [inline bfsaz/bfsaz.go]: This (replacing underscores/dashes) is coded in bfs.NormMetadata - code duplication?

Comment (mxmCherry) [reply bfsaz/bfsaz.go]: Or that's intended - no need to MIME-canonicalise keys?

Comment (dim) [reply bfsaz/bfsaz.go]: they are canonicalised in the next step

File: bfsaz/bfsaz.go [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &bfs.MetaInfo{
+		Name:        name,
+		Size:        resp.ContentLength(),
+		ModTime:     resp.LastModified(),
+		ContentType: resp.ContentType(),
+		Metadata:    bfs.NormMetadata(transKeys(resp.NewMetadata(), "_", "-")),
```
Comment (dim) [inline bfsaz/bfsaz.go]: @mxmCherry `NormMetadata(transKeys(...))`

File: bfsaz/bfsaz.go [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+		}
+		defer file.Close()
+
+		// Upload file
+		_, err = w.blob.Upload(
+			w.ctx, file,
+			azblob.BlobHTTPHeaders{
+				ContentType: w.opts.GetContentType(),
+			},
+			azblob.Metadata(transKeys(w.opts.GetMetadata(), "-", "_")),
```
Comment (dim) [inline bfsaz/bfsaz.go]: @mxmCherry same again: `GetMetadata()` copies the map and canonicalises the keys, then transKeys is applied on top

**Pull request merged**